### PR TITLE
[Decoding] Support repetition penalty

### PR DESCRIPTION
**Pull request by yzh119**
The repetition penalty (introduced in [CTRL](https://arxiv.org/abs/1909.05858)) can help prevent the LLM from generating repetitive tokens.
This PR implements the repetition penalty.

# Note
Previous the logits softmax is performed on GPU, this PR moves it to CPU to accommodate the repetition penalty.

### CODE REVIEW

**Review by MasterJH5574 — APPROVED** [review]

**Comment by MasterJH5574** [comment]
Will likely need to bring some computation to GPU side as a future enhancement.
